### PR TITLE
ESRI-235 - Proactive insights - Spam protection

### PR DESCRIPTION
**Pull request by AleksaMilivojevicCycloMedia**
ESRI-235 - Proactive insights - Spam protection

### CODE REVIEW

**Review by martin-sven-kovacic-valcon — APPROVED** [review]

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 1]
```diff
@@ -15,10 +23,18 @@ public enum EventLogLevel : short
 
   public static class EventLog
   {
+    private static int LogsLimit = LogData.Instance.LogLimit;
+    private static TimeUnit timeUnit = LogData.Instance.TimeUnit;
+    private static int logCount = 0;
+    private static DateTime lastReset;
+    private static LogData _logData = LogData.Instance;
+    private static bool isLimitReachedBefore = false;
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: inconsistent coding styles in a single class.

Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: Reached Before what?

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 2]
```diff
@@ -15,10 +23,18 @@ public enum EventLogLevel : short
 
   public static class EventLog
   {
+    private static int LogsLimit = LogData.Instance.LogLimit;
+    private static TimeUnit timeUnit = LogData.Instance.TimeUnit;
+    private static int logCount = 0;
+    private static DateTime lastReset;
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: Naming - last reset of what?

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 3]
```diff
@@ -33,22 +49,69 @@ public static IDisposable InitializeSentry(string sentryDsnUrl)
         return null;
       }
     }
+    static EventLog()
+    {
+      _logData = LogData.Instance;
+      if(_logData.LogCount >= LogsLimit)
+      {
+        isLimitReachedBefore = true;
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: isLimitReachedBefore = _logData.LogCount >= LogsLimit;

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 4]
```diff
@@ -33,22 +49,69 @@ public static IDisposable InitializeSentry(string sentryDsnUrl)
...
         return;
       }
+      ResetCounterIfNeeded();
 
       SentryLevel loggingLevel = MapEventLogTypeToSentryLevel(type);
       if (loggingLevel == SentryLevel.Error)
       {
+        if (logCount >= LogsLimit)
+        {
+          if (!isLimitReachedBefore)
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: Remove nested ifs

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 5]
```diff
@@ -15,10 +23,18 @@ public enum EventLogLevel : short
 
   public static class EventLog
   {
+    private static int LogsLimit = LogData.Instance.LogLimit;
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: Do we need to ensure in variable names that it is related to error logs only?

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+  public string SentryDsnUrl
+  {
+    get
+    {
+      return "https://96100a2b9133a1f056bf8df35d72d201@o4508125690593280.ingest.de.sentry.io/4508125706387536"; // TODO: change to proper production key
+    }
+  }
+
+  private static string LogDataFileName => Path.Combine(FileUtils.FileDir, "logdata.xml");
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs]: All other file names in this folder have Pascal Case naming.

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    {
+      FileStream streamFile = new FileStream(LogDataFileName, FileMode.Create);
+      XmlSerializer serializer = new XmlSerializer(typeof(LogData));
+      serializer.Serialize(streamFile, this);
+      streamFile.Close();
+    }
+
+    catch
+    {
+      EventLog.Write(EventLogLevel.Error, $"Street Smart: (EventLog.cs) (SaveLogData)");
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs]: Is this a circular dependency?

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        {
+          XmlSerializer serializer = new XmlSerializer(typeof(LogData));
+          _logData = (LogData)serializer.Deserialize(streamFile);
+          streamFile.Close();
+        }
+      }
+    }
+    catch
+    {
+      EventLog.Write(EventLogLevel.Error, $"Street Smart: (EventLog.cs) (LoadLogData)");
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs]: Is this a circular dependency?

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  private static LogData _logData;
+
+  public LogData() { }
+
+  public LogData(int logCount, DateTime lastResetTime)
+  {
+    LogCount = logCount;
+    LastResetTime = lastResetTime;
+  }
+  public LogData(int logCount, DateTime lastResetTime, bool useSentryLogging, int logLimit, TimeUnit timeUnit)
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs]: Why do we need two constructors?

**Review by bervacanin-cy — CHANGES_REQUESTED** [review]

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs [hunk 10]
```diff
@@ -33,22 +49,69 @@ public static IDisposable InitializeSentry(string sentryDsnUrl)
...
+      if (TimeElapsedSinceLastRestart(timeUnit) >= 1)
+      {
+        logCount = 0;
+        lastReset = DateTime.Now;
+        isLimitReachedBefore = false;
+        _logData.LogCount = logCount;
+        _logData.LastResetTime = lastReset;
+        _logData.Save();
+      }
+    }
```
Comment (martin-sven-kovacic-valcon) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/EventLog.cs]: this needs to be completely redone

whole Write method
whole ResetCounterIfNeeded method

**Review by martin-sven-kovacic-valcon — CHANGES_REQUESTED** [review]

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs (hunk 9 above)
Comment (AleksaMilivojevicCycloMedia) [reply client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Logging/LogData.cs]: We don't need two constructors here

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/AddIns/Modules/StreetSmart.cs [hunk 11]
```diff
@@ -101,6 +100,13 @@ public StreetSmart()
       MapClosedEvent.Subscribe(OnMapClosedDocument);
       ActiveMapViewChangedEvent.Subscribe(OnActiveMapViewChanged);
       ApplicationStartupEvent.Subscribe(OnApplicationStartupEvent);
+      ApplicationClosingEvent.Subscribe(OnApplicationClosingEvent);
+    }
+
+    private Task OnApplicationClosingEvent(CancelEventArgs args)
+    {
+      LogData.Instance.Save();
+      return Task.FromResult(0);
```
Comment (martin-sven-kovacic-valcon) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/AddIns/Modules/StreetSmart.cs]: return Task.CompletedTask

**Review by martin-sven-kovacic-valcon — APPROVED** [review]

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Configuration/File/LogData.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+      }
+    }
+
+    private static LogData Create()
+    {
+      var result = new LogData
+      {
+        LogCount = 0,
+        LastResetTime = DateTime.Now,
+        UseSentryLogging = true,
```
Comment (bervacanin-cy) [inline client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Configuration/File/LogData.cs]: I think we should avoid having Sentry enabled by default. We still do not have a company account.

**Review by bervacanin-cy — CHANGES_REQUESTED** [review]

File: client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Configuration/File/LogData.cs (hunk 12 above)
Comment (AleksaMilivojevicCycloMedia) [reply client-side/C#/StreetSmartArcGISPro/StreetSmartArcGISPro/Configuration/File/LogData.cs]: I agree, i will change it to disabled

**Review by bervacanin-cy — APPROVED** [review]

**Pull request merged**